src/Commands.hs

```diff
@@ -519,47 +519,38 @@ commandMacroLog ctx msgs = do
 
 commandEq :: BinaryCommandCallback
 commandEq ctx a b =
-  pure $ case cmp (a, b) of
-    Left (a', b') -> evalError ctx ("Can't compare " ++ pretty a' ++ " with " ++ pretty b') (xobjInfo a')
-    Right b' -> (ctx, Right (boolToXObj b'))
+  pure (ctx, Right (boolToXObj (cmp (a, b))))
   where
     cmp (XObj (Num aTy aNum) _ _, XObj (Num bTy bNum) _ _)
       | aTy == bTy =
-        Right $ aNum == bNum
-    cmp (XObj (Str sa) _ _, XObj (Str sb) _ _) = Right $ sa == sb
-    cmp (XObj (Chr ca) _ _, XObj (Chr cb) _ _) = Right $ ca == cb
-    cmp (XObj (Sym sa _) _ _, XObj (Sym sb _) _ _) = Right $ sa == sb
-    cmp (XObj (Bol xa) _ _, XObj (Bol xb) _ _) = Right $ xa == xb
-    cmp (XObj Def _ _, XObj Def _ _) = Right True
-    cmp (XObj Do _ _, XObj Do _ _) = Right True
-    cmp (XObj Let _ _, XObj Let _ _) = Right True
-    cmp (XObj While _ _, XObj While _ _) = Right True
-    cmp (XObj Break _ _, XObj Break _ _) = Right True
-    cmp (XObj If _ _, XObj If _ _) = Right True
-    cmp (XObj With _ _, XObj With _ _) = Right True
-    cmp (XObj MetaStub _ _, XObj MetaStub _ _) = Right True
-    cmp (XObj Address _ _, XObj Address _ _) = Right True
-    cmp (XObj SetBang _ _, XObj SetBang _ _) = Right True
-    cmp (XObj Macro _ _, XObj Macro _ _) = Right True
-    cmp (XObj Dynamic _ _, XObj Dynamic _ _) = Right True
-    cmp (XObj DefDynamic _ _, XObj DefDynamic _ _) = Right True
-    cmp (XObj The _ _, XObj The _ _) = Right True
-    cmp (XObj Ref _ _, XObj Ref _ _) = Right True
-    cmp (XObj Deref _ _, XObj Deref _ _) = Right True
-    cmp (XObj (Lst []) _ _, XObj (Lst []) _ _) = Right True
+        aNum == bNum
+    cmp (XObj (Str sa) _ _, XObj (Str sb) _ _) = sa == sb
+    cmp (XObj (Chr ca) _ _, XObj (Chr cb) _ _) = ca == cb
```

Comment:
Are all these explicit matches on XObjs needed? 

Couldn't this be simplified to:

```
cmp x x = obj x = obj x
```

Retaining the clauses for lists and arrays since they require length comparisons etc.

Reply:
Huh... Actually that seems to work fine. I also had to leave the case for symbols, since we don’t care about a symbol’s mode when checking for equality. That simplifies the function tremendously!